### PR TITLE
Add support for the awsQueryCompatible trait

### PR DESCRIPTION
**Pull request by ysaito1001**
## Motivation and Context
Implements https://github.com/awslabs/smithy-rs/issues/1872

## Description
This PR adds support for the [awsQueryCompatible](https://smithy.io/2.0/aws/protocols/aws-query-protocol.html?#aws-protocols-awsquerycompatible-trait) trait to the Rust SDK. The trait is specifically used with the `awsJson1_0` trait (`awsQueryCompatible`'s Trait selector is `service [trait|awsJson1_0]`) and allows services to continue supporting a custom error code via the [awsQueryError](https://awslabs.github.io/smithy/2.0/aws/protocols/aws-query-protocol.html#aws-protocols-awsqueryerror-trait) trait when they switch from `awsQuery` to `awsJson1_0` annotated with `awsQueryCompatible`.

When a service makes that protocol switch, the service will then respond with an additional header `x-amzn-query-error` whose value is in the form of `<error code>;<error type>`. An example looks something like
```
HTTP/1.1 400
x-amzn-query-error: AWS.SimpleQueueService.NonExistentQueue;Sender
Date: Wed, 08 Sep 2021 23:46:52 GMT
Content-Type: application/x-amz-json-1.0
Content-Length: 163

{
    "__type": "com.amazonaws.sqs#QueueDoesNotExist",
    "message": "some user-visible message"
}
```
(`AWS.SimpleQueueService.NonExistentQueue` corresponds to `error code` and `Sender` to `error type`)

The goal of this PR is to allow customers to have access to `error code` and `error type`, given a response like above (see [CHANGE.next.toml](https://github.com/awslabs/smithy-rs/pull/2398/commits/fdd98b1403e1d345b58025be9e98ee96e026074e) for more details). The code changes, therefore, include
- Check whether a service is annotated with both `awsJson1_0` annotated with `awsQueryCompatible`.
- If so, create a new class for `AwsQueryCompatible` to generate code for parsing a response to extract `error code` and `error type` from the custom header `x-amzn-query-error`.
- Make `error code` and `error type` available in `ErrorMetadata`.

## Testing
- [x] Added unit tests in `rust-runtime/inlineable/src/aws_query_compatible_errors.rs`
- [x] Added unit tests in `AwsQueryCompatibleTest.kt`
  - [x] Validated SDK can parse error code and error type fields
  - [x] Validated SDK can handle missing custom header (hence there is no `<error code>;<error type>`), still being able to obtain error code from the JSON payload instead.
- [x] Passed tests in CI

## Checklist

- [x] I have updated `CHANGELOG.next.toml` if I made changes to the smithy-rs codegen or runtime crates

----

_By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice._

### CODE REVIEW

File: CHANGELOG.next.toml [hunk 1]
```diff
@@ -235,3 +235,45 @@ message = "Support for constraint traits on member shapes (constraint trait prec
 references = ["smithy-rs#1969"]
 meta = { "breaking" = false, "tada" = true, "bug" = false, "target" = "server" }
 author = "drganjoo"
+
+[[smithy-rs]]
+message = """
+Add support for the `awsQueryCompatible` trait. This allows services to continue supporting a custom error code (via the `awsQueryError` trait) when the services migrate their protocol from `awsQuery` to `awsJson1_0` annotated with `awsQueryCompatible`. After the migration, services will include an additional header `x-amzn-query-error` in their responses whose value is in the form of `<error code>;<error type>`. An example response looks something like
```
Comment (jdisanti) [inline CHANGELOG.next.toml:241]: Can you put everything after the first sentence into an expand/collapse?

File: CHANGELOG.next.toml [hunk 2]
```diff
@@ -235,3 +235,45 @@ message = "Support for constraint traits on member shapes (constraint trait prec
...
+    }
+}
+```
+"""
+references = ["smithy-rs#2398"]
+meta = { "breaking" = false, "tada" = true, "bug" = false }
+author = "ysaito1001"
+
+[[aws-sdk-rust]]
+message = "Add support for the `awsQueryCompatible` trait, allowing the Rust SDK to migrate from the AWS Query protocol to the AWS JSON 1.0 protocol to continue supporting custom error codes."
```
Comment (jdisanti) [inline CHANGELOG.next.toml:276]: I don't think we should put this in the SDK changelog.

File: codegen-core/src/main/kotlin/software/amazon/smithy/rust/codegen/core/smithy/protocols/AwsQueryCompatible.kt [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+class AwsQueryCompatible(
+    val codegenContext: CodegenContext,
+    private val awsJson: AwsJson,
+) : Protocol {
+    private val runtimeConfig = codegenContext.runtimeConfig
+    private val errorScope = arrayOf(
+        "Bytes" to RuntimeType.Bytes,
+        "ErrorMetadataBuilder" to RuntimeType.errorMetadataBuilder(runtimeConfig),
+        "HeaderMap" to RuntimeType.Http.resolve("HeaderMap"),
+        "HeaderValue" to RuntimeType.Http.resolve("HeaderValue"),
```
Comment (jdisanti) [inline codegen-core/src/main/kotlin/software/amazon/smithy/rust/codegen/core/smithy/protocols/AwsQueryCompatible.kt:54]: Unused?
```suggestion
```

File: rust-runtime/inlineable/src/aws_query_compatible_errors.rs [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+        response.headers_mut().insert(
+            X_AMZN_QUERY_ERROR,
+            http::HeaderValue::from_static("AWS.SimpleQueueService.NonExistentQueue;Sender"),
+        );
+
+        let actual = aws_query_compatible_error_from_header(response.headers()).unwrap();
+
+        assert_eq!(
+            actual,
+            Some("AWS.SimpleQueueService.NonExistentQueue;Sender")
```
Comment (jdisanti) [inline rust-runtime/inlineable/src/aws_query_compatible_errors.rs:58]: Nit: It's nice to have a consistent order on assertions. We've been using `expected, actual` order for the most part.

**Review by jdisanti — APPROVED** [review]
LGTM! Just some minor comments.

File: CHANGELOG.next.toml (hunk 1 above)
Comment (ysaito1001) [reply CHANGELOG.next.toml]: Good suggestion! Addressed in a8386bd.

File: CHANGELOG.next.toml (hunk 2 above)
Comment (ysaito1001) [reply CHANGELOG.next.toml]: True, the change is really for `[[smithy-rs]]`. Removed `[[aws-sdk-rust]]` in a8386bd.

File: codegen-core/src/main/kotlin/software/amazon/smithy/rust/codegen/core/smithy/protocols/AwsQueryCompatible.kt (hunk 3 above)
Comment (ysaito1001) [reply codegen-core/src/main/kotlin/software/amazon/smithy/rust/codegen/core/smithy/protocols/AwsQueryCompatible.kt]: Yes. Made changes in dd1c312 without realizing `Suggested change`. Thank you though.

File: rust-runtime/inlineable/src/aws_query_compatible_errors.rs (hunk 4 above)
Comment (ysaito1001) [reply rust-runtime/inlineable/src/aws_query_compatible_errors.rs]: Thanks, I'll stick to `expected, actual` in the future. I might have done `actual, expected` in the past but at least fixed occurrences in this PR in befa41c.

**Review by jdisanti — APPROVED** [review]

**Pull request merged**